Add solution for day 5 (#36)
scalacenter/scala-advent-of-code: docs/puzzles/day5.md

sidebar_position: 5
---

by [@tgodzik](https://github.com/tgodzik)

> 

Reviewer: Maybe you need to move that after the title `# Day 5: Hydrothermal Venture`.

Author: Done!

// vent is vertical
else if vent.start.y == vent.end.y then
  for px <- rangex do update(Point(px, vent.start.y))
```

and add the end just count them all up:

```scala
map.count { case (_, v) => v > 1 }
```

Altogether the solution looks like this:

```scala
def findDangerousPoints(vents: Seq[Vent]) =
  val map = mutable.Map[Point, Int]().withDefaultValue(0)
  def update(p: Point) =
    val current = map(p)
    map.update(p, current + 1)

  for vent <- vents do
    def rangex =
      val stepx = if vent.end.x > vent.start.x then 1 else -1
      vent.start.x.to(vent.end.x, stepx)
    def rangey =
      val stepy = if vent.end.y > vent.start.y then 1 else -1
      vent.start.y.to(vent.end.y, stepy)
    // vent is horizontal
    if vent.start.x == vent.end.x then
      for py <- rangey do update(Point(vent.start.x, py))
    // vent is vertical
    else if vent.start.y == vent.end.y
      for px <- rangex do update(Point(px, vent.start.y))
  end for

  map.count { case (_, v) => v > 1 }
end findDangerousPoints
```

### Hydrothermal Venture 2: The Diagonal Menace

For the second part of the puzzle we need to take into account also the diagnal

> vents. Fortunately, they can only have an angle of 45 degrees, which means when
> iterating over ranges both x and y positions increment by 1. So we can add

Reviewer: I find this sentence hard to read:
```suggestion
vents. Fortunately, they can only have an angle of 45 degrees, which means that
both x and y positions increment by 1 at each step of the range. So we can add
```

Author: I was writing it a bit in a hurry :sweat:
